src/units/zone_loader.rs

```diff
@@ -161,50 +141,37 @@ impl ZoneLoader {
                 return Err(Terminated);
             }
 
-            Some(ApplicationCommand::RegisterZone { register }) => {
-                let res = match register.source {
-                    ZoneSource::Zonefile {
-                        path, /* Lacks XFR out settings */
-                    } => {
-                        Self::register_primary_zone(
-                            register.name.clone(),
-                            &path.to_string(),
-                            tsig_key_store,
-                            None,
-                            &self.xfr_out,
-                            &zone_updated_tx,
-                        )
-                        .await
-                    }
-                    ZoneSource::Server {
-                        addr, /* Lacks TSIG key name */
-                    } => {
-                        // Use any existing XFR inbound ACL that has been
-                        // defined for this zone from this source.
-                        Self::register_secondary_zone(
-                            register.name.clone(),
-                            tsig_key_store,
-                            Some(addr),
-                            &self.xfr_in,
-                            zone_updated_tx.clone(),
-                        )
-                    }
+            Some(ApplicationCommand::Changed(Change::ZoneSourceChanged(name, source))) => {
+                // Just remove and re-insert the zone.
+                let id = ZoneId {
+                    name: name.clone(),
+                    class: Class::IN,
                 };
+                zone_maintainer.remove_zone(id).await;
+
+                let zone = match source {
+                    ZoneLoadSource::None => return Ok(()),
 
-                match res {
-                    Err(_) => {
-                        error!("[ZL]: Error: Failed to register zone '{}'", register.name);
+                    ZoneLoadSource::Zonefile { path } => {
+                        Self::register_primary_zone(name.clone(), &path, &zone_updated_tx).await?
                     }
 
-                    Ok(zone) => {
-                        if let Err(err) = zone_maintainer.insert_zone(zone).await {
-                            error!(
-                                "[ZL]: Error: Failed to insert zone '{}': {err}",
-                                register.name
-                            );
-                        }
+                    ZoneLoadSource::Server { addr, tsig_key: _ } => {
+                        Self::register_secondary_zone(name.clone(), addr, zone_updated_tx)?
                     }
-                }
+                };
+
+                // TODO: Handle (or iron out) potential errors here.
```

Comment:
While an error shouldn't happen here, as it can only happen if inserting a zone that already exists and you just removed the zone, maybe it's worth logging something here if it happens as it is really not expected, AND would mean potentially that the zone wasn't added after all?